sympy/functions/combinatorial/numbers.py

```diff
@@ -996,6 +997,163 @@ def _eval_evalf(self, prec):
+     Fermat test. Let a be a random number between 2 and n - 1, where n is the number whose primality we are testing.
+     Then, n is probably prime if the following equation holds:
+     an mod n = a.If a number passes the Fermat test several times, then it is prime with a high probability.
+     Unfortunately, there is bad news. Certain composite numbers (non-primes) still pass the Fermat test with every number smaller than themselves.
+     These numbers are called Carmichael numbers.
+     
+     A Carmichael number will pass a Fermat primality test to every base b relatively prime to the number,
+     even though it is not actually prime. This makes tests based on Fermat's Little Theorem less effective than 
+     strong probable prime tests such as the Baillie-PSW primality test and the Miller–Rabin primality test.
+     
+     mr functions given in sympy/sympy/ntheory/primetest.py will go wrong for each and every carmichael number
+     
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Carmichael_number
+ 
+    Examples
+    ========
+    from sympy.ntheory.primetest import mr
+    if __name__ == "__main__":
+
+    # test code
+        numCarmichaels = 10
+        print "First " + str(numCarmichaels) + " Carmichael Numbers:"
+        carmichaelSet = find_first_n_carmichaels(numCarmichaels)
+        carmichaelSet = sorted(carmichaelSet)
+
+        print str(carmichaelSet)
+        print "\n---Miller-Rabin Prime Tests---"
+        print '%10s' % "Carmichael" + \
+              '%6s' % " Miller Rabin" +\
+              '%6s' % " Truth"
+        for c in carmichaelSet:
+            print '%10s' % str(c) + "|" + \
+                  '%6s' % str(mr(c, [10,2,3,9,17,8])) + "|" + \
+                  '%6s' % str(is_prime(c))
+
+        start = 100
+        end = 10000
+        print "\nCarmichael numbers between " + str(start) + " and " + str(end) + ":"
+        print str(find_carmichaels_in_range(start, end))
+    
+     Output
+    ========
+     >>> 
+    ======================== RESTART: C:/Python27/sympycarmichael.py ========================
+    First 10 Carmichael Numbers:
+    1) 561
+    2) 1105
+    3) 1729
+    4) 2465
+    5) 2821
+    6) 6601
+    7) 8911
+    8) 10585
+    9) 15841
+    10) 29341
+    [561, 1105, 1729, 2465, 2821, 6601, 8911, 10585, 15841, 29341]
+
+    ---Fermat Prime Tests---
+    Carmichael Miller Rabin Truth
+           561| False
+          1105| False
+          1729| False
+          2465| False
+          2821| False
+          6601| False
+          8911| False
+         10585| False
+         15841| False
+         29341| False
+
+    Carmichael numbers between 100 and 10000:
+    Set([128, 512, 8192, 2821, 2048, 1729, 6601, 8911, 1105, 561, 2465])
+
+    """
+    knownPrimes = Set()
+    # returns true if n is a perfect square
+    def is_perfect_square(n):
```

Comment:
This function is already implemented in ntheory/primetest.py, as `is_square`. You can import it instead of defining it again.

Reply:
ok

Reply:
thanks for advice
done now !